core/winnt/src/TWinNTSystem.cxx

```diff
@@ -2256,20 +2256,44 @@ const char *TWinNTSystem::TempDirectory() const
 /// Create a secure temporary file by appending a unique
 /// 6 letter string to base. The file will be created in
 /// a standard (system) directory or in the directory
-/// provided in dir. The full filename is returned in base
+/// provided in dir. Optionally one can provide suffix
+/// append to the final name - like extension ".txt" or ".html".
+/// The full filename is returned in base
 /// and a filepointer is returned for safely writing to the file
 /// (this avoids certain security problems). Returns 0 in case
 /// of error.
 
-FILE *TWinNTSystem::TempFileName(TString &base, const char *dir)
+FILE *TWinNTSystem::TempFileName(TString &base, const char *dir, const char *suffix)
 {
    char tmpName[MAX_PATH];
 
-   ::GetTempFileName(dir ? dir : TempDirectory(), base.Data(), 0, tmpName);
+   auto res = ::GetTempFileName(dir ? dir : TempDirectory(), base.Data(), 0, tmpName);
+   if (res == 0) {
+      ::SysError("TempFileName", "Fail to generate temporary file name");
+      return nullptr;
+   }
+
    base = tmpName;
-   FILE *fp = fopen(tmpName, "w+");
+   if (suffix && *suffix) {
+      base.Append(suffix);
+
+      if (!AccessPathName(base, kFileExists)) {
+         ::SysError("TempFileName", "Temporary file %s already exists", base.Data());
+         Unlink(tmpName);
+         return nullptr;
+      }
+
+      auto res2 = Rename(tmpName, base.Data());
+      if (res2 != 0) {
+         ::SysError("TempFileName", "Fail to rename temporary file to %s", base.Data());
+         Unlink(tmpName);
+         return nullptr;
+      }
```

Comment:
I'm not a huge fan, because here we still have a race between `AccessPathName` and `Rename`.

What would speak against the `CreateFile` with `CREATE_NEW` and `_open_osfhandle` approach?

Reply:
@jblomer 

Can you provide code which I can insert here?
I guess, you want to replace `Rename` function here?

Reply:
I'll check with @bellenot later today

Reply:
I just tried this approach and it seems to work, but I'm still performing some tests to make sure it works in all cases. Stay tuned!

Reply:
Another solution would be to use the `GetTempPath` and `CoCreateGuid` to generate a `UUID` for the file. But then this code:
```
   std::string base = ".fileAsyncOpen_";
   FILE *f = gSystem->TempFileName(base, nullptr, ".html");
```
generates:
```
C:\Users\bellenot\AppData\Local\Temp\.fileAsyncOpen_86BE05A0-104A-4E2A-A3D7-E2CF1EF72EB5.html
```
Otherwise the current implementation works, beside the fact that `GetTempFileName()` only uses up to the first three characters of the `base` string